videosink/writer.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package videosink
+
+import (
+	"bytes"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"net/textproto"
+	"strconv"
+)
+
+// randomBoundary generates a MIME multipart boundary compatible with RFC 2046
+// (section 5.1.1).
+func randomBoundary() string {
+	var buf [34]byte
+	if _, err := io.ReadFull(rand.Reader, buf[:]); err != nil {
+		panic(err)
+	}
+	return fmt.Sprintf("%x", buf[:])
+}
+
+type partWriter struct {
+	u        io.Writer
+	boundary string
+	started  bool
+}
+
+func makePartWriter(u io.Writer) partWriter {
+	return partWriter{
+		u:        u,
+		boundary: randomBoundary(),
+	}
+}
+
+// writeFrame sends a single part of a MIME multipart entity, ensuring it's
+// fully written by the time the function returns.
+//
+// The caller-owned headers are modified to set a Content-Length header.
+//
+// Go has a writer for MIME multipart messages in "mime/multipart".Writer. As
+// of Go 1.17 it's not suitable for writing a neverending stream of parts where
+// each must be flushed to the client with the part-ending boundary line.
+func (w *partWriter) writeFrame(header textproto.MIMEHeader, body []byte) error {
+	header.Set("Content-Length", strconv.FormatInt(int64(len(body)), 10))
+
+	var buf bytes.Buffer
+
+	if !w.started {
```

Comment:
Why do you create a temporary buffer, to reduce the number of write calls to the response?

Reply:
This function was heavily inspired by Go's own MIME multipart writer:

https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/mime/multipart/writer.go;l=97-130;drc=a6293995c55659f51e0662e7656f395633c99b5b

Yes, it does reduce write calls. Looking at it now we could also use the buffer pool here to avoid an allocation (though I'd first check if Go's escape analysis didn't eliminate that already).

Are you concerned about something in specific?

Reply:
No, just curious.